### ISSUE TITLE
Did something change?

### ISSUE THREAD

**Issue by frozenpandaman** [drunohazarb/4chan-captcha-solver#1]
As of yesterday it's now quite bad at solving, almost always missing characters. I'd say it got it right about 80-90% of the time before and now it's around ~~10-20%~~ 5% (at most).

EDIT: ok i know the answer is "yes something changed" so this is really "will you (or someone) fix it please & thank you"

**Comment by drunohazarb** [comment]
Indeed. 4chan changed their captcha and the letters are now noticeably smaller, breaking most captcha solvers out there. I'm looking into it.

**Comment by HairyMilkshakes** [comment]
New captcha samples, a few thousand of them to help you with the new model.
[1m831a.zip](https://github.com/drunohazarb/4chan-captcha-solver/files/12252795/1m831a.zip)

**Comment by drunohazarb** [comment]
I have implemented the workaround used by JKCS (scaling the captcha) to make the solver a little more bearable to use while I figure out how to train the new model
0ad7f6f312bd0d3be132b8ded5bf9e4c6c3923a5

**Comment by frozenpandaman** [comment]
Working pretty well! Thanks for this! (Feel free to close the issue if you consider it more-or-less 'resolved' since I think this improves usability a lot.)

**Comment by MooseZM** [comment]
There was an improvement in accuracy but the new captcha has 6 characters, this script only outputs 5 (those 5 are usually correct but it skips one character randomly)

**Comment by coomdev** [comment]
@drunohazarb I'm JKCS' dev. I'm also working on training a new model (though I don't know much about what I'm doing), I managed to get some pretty promising results by training on the dataset from the zip published by Chance's, but I need many more captchas to make it at least as accurate as JKCS used to be.

I wrote this userscript to collect captchas, for 4chanX users since one person on /g/ said they'd be interested in contributing.

https://gist.github.com/coomdev/47d1b243a53c3690dcff14a6c01b9267/raw/dba569d8a93f0a8806c085e8a941a1b4d7e189f1/cc.user.js

**Comment by frozenpandaman** [comment]
Looks like others are collecting captchas too. I got a prompt within the KurobaEx app which links this issue:
https://github.com/moffatman/chan/issues/144

**Comment by JonseyJones** [comment]
Quote:
"the answer is I likely won't work on it completely 
i can share generation and training scripts with you
or if you are allergic to ML I can share just the generation script and you'd update it to produce captches that look like current 4chan's
or write your own
I then would use your script tp generate  a dataset of captches and train on it
the one I made is written in perl" https://cdn.discordapp.com/attachments/1137092744741929051/1137093467374366720/generate.7z
this is an example output needed to train the neural network
https://cdn.discordapp.com/attachments/1137092744741929051/1137093824489984142/x.7z
i use calamari ocr tpo train but again if you are not in capacity of doing that, I can train myself"
"generate 50k captchas, train a neural network from scratch for it, run my script that converts the trained neural network into tensorflow-lite file, convert that file into base64, and copy-paste that into the userscript"
From our man Automatic. Hope it helps, you guys can contact him on discord, see the link on his github.

**Comment by moffatman** [comment]
Now that kuroba-ex has added the option, I'm getting a lot more captchas. Will post once I have 10k in the next few days.

**Comment by moffatman** [comment]
You can grab the first 10k here. It does include the ~1.5k I shared previously.

https://captcha.chance.surf/images_10k.zip

I didn't check for any misspelled letters, sometimes people will type a similar looking letter, which is still accepted. But to train correctly, they should be renamed as follows:

```dart
const remap = {
  'B': '8',
  'F': 'P',
  'U': 'V',
  'Z': '2',
  'O': '0'
};
```

**Comment by coomdev** [comment]
Ok, I'm starting a training on a 75:25 split, if it looks good (and for now, metrics show it's bretty good) I'll release the notebook on JKCS and a beta model (since you'll need to implement both the image preprocessing and maybe inference) do a training on the full dataset and release a hopefully stable and complete model.
I hope my laptop won't melt by then, it's smelling things I've never smelt before.

**Comment by coomdev** [comment]
![image](https://github.com/drunohazarb/4chan-captcha-solver/assets/73244916/7c53ac1a-1371-476c-9d6b-faae4a4f2da0)
looks bretty good, it fails with at most one letter, and 3/4 of those fails are things that the pre-process failed to properly cleanup.
I'll publish the models on [JKCS's](https://git.coom.tech/araragi/JKCS) repo with the jupyter notebook, go to sleep and continue when I wake up with cleaning up and full training, but in the meanwhile you can pull and evaluate the model and figure out how to integrate it in your apps.

**Comment by moffatman** [comment]
Do you have any estimate the % fully correct? I found on my own solver preprocessing is quite difficult now, the letters and noise are a lot closer in size. Just from that image -- 75%?

**Comment by coomdev** [comment]
Yeah, I just ran against the entire 2514 validation set, got 1916 correct (76%)

**Comment by Yukariin** [comment]
Fixed the model and trained it on combined dataset (10k from @moffatman + ~500 from @coomdev):
![image](https://github.com/drunohazarb/4chan-captcha-solver/assets/6450725/ace0dae4-94cd-4e8f-a3ff-d212663f567a)

[captcha_75_25.h5](https://files.catbox.moe/idtjid.h5)
[captcha_75_25.keras](https://files.catbox.moe/idtjid.h5)
[captcha_75_25.tf](https://files.catbox.moe/vfj2zd.zip)
Updated training notebook [here](https://files.catbox.moe/5hvq0j.ipynb)

**Comment by moffatman** [comment]
So what is the accuracy now?

**Comment by Yukariin** [comment]
It solved 2529 of 2653 val images, so ~95%

**Comment by coomdev** [comment]
Woah, impressive.

**Comment by coomdev** [comment]
My bad. My exporter was shitting up because of the extension so I thought otherwise.

**Comment by moffatman** [comment]
This solver is awesome thanks @Yukariin. I wonder if any more images could be helpful? I have another 10k more now. But it is already pretty much there. Only thing I found, rarely it does confuse D and 0 (which is pretty difficult even by eye sometimes). Also I missed the mapping '5' -> 'S' when I posted before, so the model has been trained with some small number of '5's where it should be 'S'.

<img width="466" alt="image" src="https://github.com/drunohazarb/4chan-captcha-solver/assets/789275/f9917c96-2996-45bd-8043-4f962f14ff37">

**Comment by KurubaEX** [comment]
Do I have to update the kuruba app in anyway to get the better captcha solver? Or is it automatic

**Comment by HairyMilkshakes** [comment]
> Do I have to update the kuruba app in anyway to get the better captcha solver? Or is it automatic

Doesn't yours need a captcha APK as complement to autosolve?

**Comment by KurubaEX** [comment]
> > Do I have to update the kuruba app in anyway to get the better captcha solver? Or is it automatic
> 
> Doesn't yours need a captcha APK as complement to autosolve?

Yeah

> > Do I have to update the kuruba app in anyway to get the better captcha solver? Or is it automatic
> 
> Doesn't yours need a captcha APK as complement to autosolve?

Yeah. So this isn't for Android kuruba?

**Comment by Yukariin** [comment]
@moffatman more data can, probably, slightly improve model's accuracy. Especially for cases like D-0
I also noticed that in @coomdev dump the image `PHPJR.png` is stored in it's own directory named `PHPJ` with image named as just `R.png`. That's probably userscript or archiver bug.

**Comment by moffatman** [comment]
I actually had a bug in my own app UI, sometimes it showed letters as 0 instead of the proper one. After I fixed that, I only saw 1 mistake out of like 50+ tries (0 instead of A). I haven't wired it all up to run in batch yet.

**Comment by coomdev** [comment]
@KurubaEX I don't know what's kuruba, but you can't just swap the old model automatic1111 for this one, the input tensor shape is different, and there's also a preprocessing step to apply ([connected component labeling](https://git.coom.tech/araragi/JKCS/src/branch/master/src/ccl.ts) then keeping the 8 biggest components (by area), then removing components with height below 20px).

@Yukariin yeah looks like someone typed in a backslash in their captcha and it was still accepted by 4chan. My archive viewer still sees "PHPJ\R" as its own foldername, but yours may have interpreted the \ as a DOS directory separator. I'll fix this serverside, thanks for the report.

**Comment by Yukariin** [comment]
Updated [notebook](https://files.catbox.moe/5lxzqe.ipynb) - fixed 5->S, added accuracy calculation on val data
Also retrained the model, accuracy is sligthly better (95.76% vs 95.39% before):
[captcha_75_25.h5](https://files.catbox.moe/egsmqc.h5)
Noticed errors on some samples, probably caused by the pre-processing algo - it discards letter parts. For example, the firts `Y`:
![image](https://github.com/drunohazarb/4chan-captcha-solver/assets/6450725/bbddc57f-edd2-4c54-9db3-c5eacdfd0d49)

**Comment by coomdev** [comment]
Yeah, it's not perfect, because of the lines used to cut characters. I already tried to do an opening operation to work around that but it resulted in worse segmentation overall. 
Though I haven't tried doing that *after* eliminating the smaller noise...
You could also try allowing more than the 8 largest components, assuming the model can withstand more noise. But the base accuracy is so high that the chance of failing two consecutive captchas is basically non-existent, so I don't think it's worth putting much more effort into preprocessing.

**Comment by Yukariin** [comment]
Yeah not sure if it’s worth it.

**Comment by Yukariin** [comment]
Okay, it's getting really interesting
I decided to try to train the model on raw data without any pre-processing and results are surprisingly good. I got 2657/2695 on val data which is 98.59%
I'd like to ask @moffatman and @coomdev to test it on your data and in you apps/extensions:
[Notebook](https://files.catbox.moe/tx7uhq.ipynb)
[captcha_75_25.h5](https://files.catbox.moe/9qsymy.h5)
[captcha_75_25.tf](https://files.catbox.moe/wrnais.zip)

**Comment by moffatman** [comment]
I dropped my own preprocessing after the new updates, it's too hard to do it correctly. So not surprising. Will try it on a ton of "previously unseen" captchas and update.

**Comment by moffatman** [comment]
Yesterday's model (idtjid.h5)

```
Correct: 0.9201347125330768 (3825 / 4157)
```

Today's model (9qsymy.h5)

```
Correct: 0.9579023334135194 (3982 / 4157)
```

**Comment by Yukariin** [comment]
The differnce probably because of pre-processing on the first model (idtjid.h5).
Still, I'm surprized it shows 90+ accuracy on ~5k of "previously unseen" files.

**Comment by moffatman** [comment]
Actually, the problem was that I forgot to apply the substitutions to the validation labels, real results are below

Yesterday's model (idtjid.h5)

```
Correct: 0.9485205677171037 (3943 / 4157)
```

Today's model (9qsymy.h5)

```
Correct: 0.9862881885975463 (4100 / 4157)
```

Crazy good!

**Comment by coomdev** [comment]
Damn, it turns out all I did was detrimental :^(

**Comment by drunohazarb** [comment]
[80c0560](https://github.com/drunohazarb/4chan-captcha-solver/commit/80c05606f2899733a5cbad6b84162a4464c18e30)
